app/controllers/application_controller.rb

```diff
@@ -2,9 +2,126 @@
 
 class ApplicationController < Sinatra::Base
 
-  configure do
-    set :public_folder, 'public'
-    set :views, 'app/views'
-  end
+	configure do
+		set :public_folder, 'public'
+		set :views, 'app/views'
+		enable :sessions
+	end
 
-end
\ No newline at end of file
+	get '/' do
+		erb :index
+	end
+
+	get '/signup' do
+		if !session[:id].nil?
+			redirect '/tweets'
+		end
+		erb :'/users/create_user'
+	end
+
+	post '/signup' do
+		@user = User.create(params)
+		if params[:username] == "" || params[:password] == "" || params[:email] == ""
+			redirect '/signup'
+		end
+
+		if @user.save
+			session[:id] = @user.id
+			redirect '/tweets'
+		else
+			redirect "/signup"
+		end
+	end
+
+	post "/login" do
+		@user = User.find_by(username: params[:username])
+		if @user && @user.authenticate(params[:password])
+			session[:id] = @user.id
+			redirect "/tweets"
+		else
+			redirect "/failure"
+		end
+	end
+
+	get '/tweets' do
+		if session[:id].nil?
+			redirect '/login'
+		end
+		@user = User.find_by_id(session[:id])
+		erb :'tweets/tweets'
+	end
+
+	get '/login' do
+		if !session[:id].nil?
+			redirect '/tweets'
+		end
+		erb :'users/login'
+	end
+
+	get '/logout' do
+		session.clear
+		redirect '/login'
+	end
+
+	get '/users/:user_slug' do
+		@user = User.find_by_slug(params[:user_slug])		
+		erb :'/tweets/show_tweet'
+	end
+
+	get '/tweets/new' do
+		if session[:id].nil?
+			redirect '/login'
+		end		
+		erb :'/tweets/new'
+	end
+
+	post '/tweets/new' do
+		if params[:content] == ""
+			redirect :'/tweets/new'	
+		end
+		@tweet = Tweet.create(params)
+		@tweet.user_id = session[:id]
+		@tweet.save
+	end
+
+	get '/tweets/:tweet_id' do
+		if session[:id].nil?
+			redirect '/login'
+		end
+		@tweet = Tweet.find_by_id(params[:tweet_id])
+		erb :'/tweets/display_single_tweet'
+	end
+
+	get '/tweets/:tweet_id/edit' do
+		@tweet = Tweet.find_by_id(params[:tweet_id])
+		if session[:id].nil?
+			redirect '/login'
+		end
+		if session[:id] != @tweet.user_id
+			redirect '/tweets/:tweet_id'
+		end
+		erb :'/tweets/edit'
+	end
+
+	post '/tweets/:tweet_id/edit' do
+		@tweet = Tweet.find_by_id(params[:tweet_id])
+		if session[:id] != @tweet.user_id
+			redirect '/tweets/:tweet_id'
+		end
+		if params[:content] != ""
+			@tweet.content = params[:content]
+			@tweet.save
+		end
+		#redirect '/tweets/:tweet_id'
+	end
+
+	post '/tweets/:tweet_id/delete' do
+		@tweet = Tweet.find_by_id(params[:tweet_id])
+		if session[:id] != @tweet.user_id
```

Comment:
Prefer to find tweets that are already only associated to the `current_user` than to manually check the `user_id` foreign key of the tweet against the `current_user.id`

```ruby
tweet = current_user.tweets.find_by(:id => params[:tweet_id])
if tweet && tweet.destroy
  redirect "/tweets"
else
  redirect "/tweets/#{tweet.id}"
end
```

We're using the association collection method provided by [`has_many`](http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_many) to query the user's tweets. 

Note the usage of [`find_by`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find-by) vs [`find`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find). Had we used `find` here, ActiveRecord would raise an error in the event that we could not find a tweet associated with this user. `find_by`, like `find_by_id`, will allow this action to fail gracefully, simply returning nil if the user is trying to delete a tweet that they do not have access to.